linkerd/app/outbound/src/endpoint.rs

```diff
@@ -71,49 +67,152 @@ pub struct TcpEndpoint {
     fn into(self) -> SocketAddr {
         self.addr
     }
 }
 
-/// Used to resolve endpoints.
-impl Into<Option<Addr>> for &'_ TcpLogical {
-    fn into(self) -> Option<Addr> {
-        Some(self.addr.into())
+impl std::fmt::Debug for TcpLogical {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TcpLogical")
+            .field("addr", &self.addr)
+            .field(
+                "profile",
+                if self.profile.is_some() {
+                    &"Some(..)"
+                } else {
+                    &"None"
+                },
+            )
+            .finish()
     }
 }
 
-// === impl HttpAccept ===
+// === impl TcpConcrete ===
+
+impl From<(Option<Addr>, TcpLogical)> for TcpConcrete {
+    fn from((resolve, logical): (Option<Addr>, TcpLogical)) -> Self {
+        Self { resolve, logical }
+    }
+}
 
-impl From<(http::Version, TcpLogical)> for HttpAccept {
-    fn from((version, TcpLogical { addr }): (http::Version, TcpLogical)) -> Self {
+/// Used as a default destination when resolution is rejected.
+impl Into<SocketAddr> for &'_ TcpConcrete {
+    fn into(self) -> SocketAddr {
+        self.logical.addr
+    }
+}
+
+// === impl HttpLogical ===
+
+impl HttpLogical {
+    pub fn addr(&self) -> Addr {
+        self.profile
+            .as_ref()
+            .and_then(|p| p.borrow().name.clone())
+            .map(|n| Addr::from((n, self.orig_dst.port())))
+            .unwrap_or_else(|| self.orig_dst.into())
+    }
+}
+
+impl From<(http::Version, TcpLogical)> for HttpLogical {
+    fn from((version, TcpLogical { addr, profile }): (http::Version, TcpLogical)) -> Self {
         Self {
             version,
             orig_dst: addr,
+            profile,
         }
     }
 }
 
-impl Into<SocketAddr> for &'_ HttpAccept {
+/// For normalization when no host is present.
+impl Into<SocketAddr> for &'_ HttpLogical {
     fn into(self) -> SocketAddr {
         self.orig_dst
     }
 }
 
+impl Into<Addr> for &'_ HttpLogical {
+    fn into(self) -> Addr {
+        self.addr()
+    }
+}
+
+impl Into<Option<profiles::Receiver>> for &'_ HttpLogical {
+    fn into(self) -> Option<profiles::Receiver> {
+        self.profile.clone()
+    }
+}
+
+// Used to set the l5d-canonical-dst header.
+impl<'t> From<&'t HttpLogical> for http::header::HeaderValue {
+    fn from(target: &'t HttpLogical) -> Self {
+        http::header::HeaderValue::from_str(&target.addr().to_string())
+            .expect("addr must be a valid header")
+    }
+}
+
+impl std::fmt::Debug for HttpLogical {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HttpLogical")
+            .field("version", &self.version)
+            .field("orig_dst", &self.orig_dst)
+            .field(
+                "profile",
+                if self.profile.is_some() {
+                    &"Some(..)"
+                } else {
+                    &"None"
+                },
```

Comment:
as i mentioned above, maybe consider
```suggestion
                &format_args!("{}", if self.profile.is_some() {
                    "Some(..)"
                } else {
                    "None"
                }),
```